### PR TITLE
refracted tidb operator build jenkins

### PR DESCRIPTION
**Pull request by lijie0123**
Signed-off-by: lijie <lijie@pingcap.com>

### CODE REVIEW

File: jenkins/pipelines/cd/tidb-operator.groovy [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+                                    sh '''git clean -f .'''
+                                    GitHash = scmVars.GIT_COMMIT
+                                }
+                                println("GitHash: $GitHash")
+                            }
+                        }
+                        stage("bin") {
+                            steps {
+                                sh """set -eux
+                                    go env -w GOPROXY="https://goproxy.pingcap.net,direct"
```
Comment (wuhuizuo) [inline jenkins/pipelines/cd/tidb-operator.groovy:147]: prefer to using env option

File: jenkins/pipelines/cd/tidb-operator.groovy [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+                                    go mod download
+                                    git checkout -- go.sum
+                                    """
+                                sh "git status"
+                                sh "GOOS=linux GOARCH=arm64 make build"
+                                sh "GOOS=linux GOARCH=amd64 make build"
+                                // todo only x86
+                                sh "GOOS=linux GOARCH=amd64 make debug-build"
```
Comment (wuhuizuo) [inline jenkins/pipelines/cd/tidb-operator.groovy:155]: too many tiny steps

File: jenkins/pipelines/cd/tidb-operator.groovy [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+                                          rm tidb-operator.tar.gz'''
+                                    def scmVars = checkout changelog: false, poll: false, scm: [
+                                            $class           : 'GitSCM',
+                                            branches         : [[name: "${params.GitRef}"]],
+                                            userRemoteConfigs: [[
+                                                                        refspec: '+refs/heads/*:refs/remotes/origin/* +refs/pull/*:refs/remotes/origin/pull/*',
+                                                                        url    : 'https://github.com/pingcap/tidb-operator.git',
+                                                                ]]
+                                    ]
+                                    sh '''git clean -f .'''
```
Comment (wuhuizuo) [inline jenkins/pipelines/cd/tidb-operator.groovy:138]: why need this step, checkout step has a option to clean git dir

File: jenkins/pipelines/cd/tidb-operator.groovy [hunk 4]
```diff
@@ -0,0 +1,409 @@
+package cd
+
+String GitHash
+String ReleaseTag
+def EnableE2E = false
```
Comment (wuhuizuo) [inline jenkins/pipelines/cd/tidb-operator.groovy:5]: is it mutable?

File: jenkins/pipelines/cd/tidb-operator.groovy [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+      exec:
+        command: ["docker", "info"]
+      initialDelaySeconds: 10
+      failureThreshold: 6
+'''
+final goBuildYaml = '''
+apiVersion: v1
+kind: Pod
+metadata:
+  name: golang
```
Comment (wuhuizuo) [inline jenkins/pipelines/cd/tidb-operator.groovy:48]: this filed is needless

File: jenkins/pipelines/cd/tidb-operator.groovy (hunk 4 above)
Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy:6]: they are immutable

File: jenkins/pipelines/cd/tidb-operator.groovy (hunk 5 above)
Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy:50]: they will be move to a separate yaml file

File: jenkins/pipelines/cd/tidb-operator.groovy [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+                                sh "GOOS=linux GOARCH=amd64 make debug-build"
+
+                                script {
+                                    sh "mkdir -p ${TOOLS_BUILD_DIR}"
+                                    for (OS in ["linux", "darwin", "windows"]) {
+                                        for (ARCH in ["amd64", "arm64"]) {
+                                            if ("$OS/$ARCH" == "windows/arm64") {
+                                                // unsupported platform
+                                                continue
+                                            }
```
[the remark this answers is not in the document]
Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy:168]: matrix is hard to share agent, I'd use it if I can

File: jenkins/pipelines/cd/tidb-operator.groovy [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+                            }
+                        }
+                    }
+                }
+                stage("Publish Files") {
+                    agent {
+                        kubernetes {
+                            yaml uploaderYaml
+                            defaultContainer 'uploader'
+                            cloud "kubernetes-ng"
```
[the remark this answers is not in the document]
Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy:355]: cloud field?

File: jenkins/pipelines/cd/tidb-operator.groovy [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+                            environment {
+                                QINIU_BUCKET_NAME = "charts";
+                            }
+                            steps {
+                                sh "curl http://charts.pingcap.org/index.yaml -o index.yaml"
+                                container("helm") {
+                                    sh "helm repo index . --url http://charts.pingcap.org/ --merge index.yaml"
+                                }
+                                sh "cat index.yaml"
+                                sh "upload_qiniu.py index.yaml index.yaml"
```
[the remark this answers is not in the document]
Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy:392]: helm is installed in another container, cat print too long, so separate may be better

File: jenkins/pipelines/cd/tidb-operator.groovy [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+                                QINIU_BUCKET_NAME = "tidb";
+                            }
+                            steps {
+                                script {
+                                    for (OS in ["linux", "darwin", "windows"]) {
+                                        for (ARCH in ["amd64", "arm64"]) {
+                                            if ("$OS/$ARCH" == "windows/arm64") {
+                                                // unsupported platform
+                                                continue
+                                            }
```
[the remark this answers is not in the document]
Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy:406]: the same situation

File: jenkins/pipelines/cd/tidb-operator.groovy [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+                                            sh "regctl image copy hub.pingcap.net/rc/${component}:${ReleaseTag}  registry.cn-beijing.aliyuncs.com/tidb/${component}:${ReleaseTag}"
+                                        }
+                                    }
+                                    stage("dockerhub") {
+                                        environment { HUB = credentials('dockerhub-pingcap') }
+                                        steps {
+                                            sh 'set +x; regctl registry login docker.io -u $HUB_USR -p $(printenv HUB_PSW)'
+                                            sh "regctl image copy hub.pingcap.net/rc/${component}:${ReleaseTag}  docker.io/pingcap/${component}:${ReleaseTag}"
+                                        }
+                                    }
```
[the remark this answers is not in the document]
Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy]: no，matrix in matrix is not supported

File: jenkins/pipelines/cd/tidb-operator.groovy (hunk 2 above)
Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy:159]: any ideas?  every step produce different production

File: jenkins/pipelines/cd/tidb-operator.groovy (hunk 3 above)
Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy:142]: go mod download may change go.sum, causing version tag contains `-dirty`

File: jenkins/pipelines/cd/tidb-operator.groovy (hunk 9 above)
[the remark this answers is not in the document]
Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy:406]: matrix can not be applied to single agent

File: jenkins/pipelines/cd/tidb-operator.groovy (hunk 5 above)
Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy]: removed

File: jenkins/pipelines/cd/tidb-operator.groovy (hunk 4 above)
Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy:6]: but they are compute in the first stage

File: jenkins/pipelines/cd/tidb-operator.groovy (hunk 10 above)
[the remark this answers is not in the document]
Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy]: and the credentian must use literature, even final var is not supported

File: jenkins/pipelines/cd/tidb-operator.groovy (hunk 7 above)
[the remark this answers is not in the document]
Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy]: done

File: jenkins/pipelines/cd/tidb-operator.groovy (hunk 4 above)
Comment (wuhuizuo) [reply jenkins/pipelines/cd/tidb-operator.groovy:6]: please using `final`

Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy:6]: they are compute in the first stage

File: jenkins/pipelines/cd/tidb-operator.groovy (hunk 2 above)
Comment (lijie0123) [reply jenkins/pipelines/cd/tidb-operator.groovy:152]: I think split may looks better, they produce different target, they are joined in one stage just for reduce time

**Pull request merged**